helios-client/src/main/java/com/spotify/helios/auth/CrtAuthProvider.java

```diff
@@ -0,0 +1,165 @@
+
+  @VisibleForTesting
+  CrtAuthProvider(final RequestDispatcher dispatcher,
+                  final String authServer,
+                  final String username,
+                  final Signer signer) {
+    this.dispatcher = dispatcher;
+    this.username = username;
+    this.crtClient = new CrtAuthClient(signer, authServer);
+  }
+
+  @Override
+  public String currentAuthorizationHeader() {
+    // TODO (dxia) Should we use a lock here? It doesn't really matter if the token isn't
+    // thread-safe. We should get valid tokens each time. The lock just prevents
+    // another thread from doing extra work to get an extra token when it could just wait.
+    if (token == null) {
+      synchronized (lock) {
+        try {
+          token = renewAuthorizationHeader(null).get();
+        } catch (InterruptedException | ExecutionException e) {
+          log.error("Failed to complete crtauth handshake.");
+          throw Throwables.propagate(e);
+        }
+      }
+    }
+
+    return token;
+  }
+
+  @Override
+  public ListenableFuture<String> renewAuthorizationHeader(final String ignored) {
+    final String authRequest = CrtAuthClient.createRequest(username);
+    try {
+      final HeliosRequest request = HeliosRequest.builder()
+          .method("GET")
+          .uri(new URI(AUTH_URI))
+          .appendHeader(CRT_HEADER, "request:" + authRequest)
+          .build();
+
+      final ListenableFuture<Response> challengeFuture = dispatcher.request(request);
+
+      final ListenableFuture<Response> tokenFuture =
+          Futures.transform(challengeFuture, new AsyncFunction<Response, Response>() {
+            @Override
+            public ListenableFuture<Response> apply(@NotNull final Response response)
+                throws Exception {
+              final String challengeHeader = response.header(CRT_HEADER);
+              // TODO (dxia) Should we check status codes or just simply check headers as I've done?
+              if (isNullOrEmpty(challengeHeader)) {
+                return immediateFailedFuture(new CrtAuthException(
+                    "Missing HTTP X-CHAP header for crtauth challenge."));
+              }
+
+              final int i = challengeHeader.indexOf(':');
+              if (i == -1) {
+                return immediateFailedFuture(new CrtAuthException(
+                    "Invalid HTTP X-CHAP header for crtauth challenge."));
+              }
+
+              final String challenge = challengeHeader.substring(i + 1);
+              final String crtResponse = crtClient.createResponse(challenge);
+              final HeliosRequest request = HeliosRequest.builder()
+                  .method("GET")
+                  .uri(new URI(AUTH_URI))
+                  .appendHeader(CRT_HEADER, "response:" + crtResponse)
+                  .build();
+
+              return dispatcher.request(request);
+            }
+          });
+
+      return Futures.transform(tokenFuture, new AsyncFunction<Response, String>() {
```

Comment:
minor but since every path through this AsyncFunction returns an immediate future, this can just be a Function (and you don't have to wrap the return values in Futures)